### PR TITLE
REST API: Sync /themes endpoint with Core's

### LINKED ISSUE #23054: Remove or update added theme fields
The [core ticket](https://core.trac.wordpress.org/ticket/49906) for adding additional fields to the Themes controller has been committed. The [Gutenberg shim](https://core.trac.wordpress.org/ticket/49906) should be removed or updated to match Core's behavior. Currently it is stomping over the Core fields as they have a different structure.

### PR DESCRIPTION
**Pull request by ockham**
## Description

In #21578, I added a few fields to Core's `/themes` endpoint, for use by the Site Editor's Template Switcher (see both #21578 and #21768). I then submitted those changes as a [PR](https://github.com/WordPress/wordpress-develop/pull/222) against Core. That PR underwent a number of modifications and was eventually merged; the new fields will be part of the `/themes` endpoint exposed by Core starting from the next WP release.

This PR updates the fields added by Gutenberg to follow the same semantics, as well as the callsites that use that endpoint.

Fixes #23054.

## How has this been tested?
Test the template switcher's theme preview as described in #21768. You need to do this both with WP 5.4, and WordPress' current `master` branch. Verify that the theme preview works in both cases.

Alternatively:
1. Check out Gutenberg's `master` branch, and make sure you're using it with WordPress' latest `master` branch.
2. Apply the following patches to disable preloading of the `/themes` endpoint, and to suppress Gutenberg's addition of fields to the `/themes` endpoint (so that we get the `/themes` endpoint to reply with the fields that are defined in Core):

```diff
diff --git a/lib/edit-site-page.php b/lib/edit-site-page.php
index 8371486663..e5a43ae479 100644
--- a/lib/edit-site-page.php
+++ b/lib/edit-site-page.php
@@ -187,7 +187,7 @@ function gutenberg_edit_site_init( $hook ) {
                '/wp/v2/types?context=edit',
                '/wp/v2/taxonomies?per_page=100&context=edit',
                '/wp/v2/pages?per_page=100&context=edit',
-               '/wp/v2/themes?status=active',
+               //'/wp/v2/themes?status=active',
                array( '/wp/v2/media', 'OPTIONS' ),
        );
        $preload_data  = array_reduce(
```

```diff
diff --git a/lib/rest-api.php b/lib/rest-api.php
index c0f1c2aa31..9d11fa23ce 100644
--- a/lib/rest-api.php
+++ b/lib/rest-api.php
@@ -126,7 +126,7 @@ function gutenberg_register_rest_block_directory() {
        $block_directory_controller = new WP_REST_Block_Directory_Controller();
        $block_directory_controller->register_routes();
 }
-add_filter( 'rest_api_init', 'gutenberg_register_rest_block_directory' );
+//add_filter( 'rest_api_init', 'gutenberg_register_rest_block_directory' );
 
 /**
  * Registers the Block types REST API routes.

```

3. Go to the Site Editor.
4. Check the Network tab for the `/themes` endpoint's response, and copy it (in JSON format) to a text editor.
5. Revert the `lib/rest-api.php` patch, but keep the `lib/edit-site-page.php` patch.
5. Switch to this branch.
6. Apply the following patch to remove all of the fields that Core's endpoint provides (except for `theme_supports`, which predates WordPress/wordpress-develop#222) in order to rely on Gutenberg to add those fields:

```
diff --git a/lib/rest-api.php b/lib/rest-api.php
index 5eccb8c3f9..a1e9409ab7 100644
--- a/lib/rest-api.php
+++ b/lib/rest-api.php
@@ -64,8 +64,9 @@ add_filter( 'rest_request_after_callbacks', 'gutenberg_filter_oembed_result', 10
  * @param WP_REST_Request  $request  Request object.
  */
 function gutenberg_filter_rest_prepare_theme( $response, $theme, $request ) {
-       $data   = $response->get_data();
-       $fields = array_keys( $data );
+       $theme_supports = $response->get_data()['theme_supports'];
+       $data           = array();
+       $fields         = array_keys( $data );
 
        /**
         * The following is basically copied from Core's WP_REST_Themes_Controller::prepare_item_for_response()
@@ -122,6 +123,8 @@ function gutenberg_filter_rest_prepare_theme( $response, $theme, $request ) {
                }
        }
 
+       $data['theme_supports'] = $theme_supports;
+
        $response->set_data( $data );
        return $response;
 }
```

7. Repeat step 4.
8. Compare the network responses -- they should be identical.

## Types of changes
Update to sync with Core.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by Addison-Stavlo — COMMENTED** [review]
I tested this on 5.4 and master, the theme preview works as expected.  Although I noticed this still works on current Gutenberg master branch, so this isn't fixing the theme preview?  That's more of a smoke test to ensure the endpoint is still working?

**Comment by ockham** [comment]
> I tested this on 5.4 and master, the theme preview works as expected. Although I noticed this still works on current Gutenberg master branch, so this isn't fixing the theme preview? That's more of a smoke test to ensure the endpoint is still working?

Uh, you're right, there was a mistake in my instructions :sweat_smile: I've updated them now, I hope they make more sense now. They're still a bit complex -- we're basically comparing the output

1. from Core's `master`, with Gutenberg's added fields suppressed
2. to Core's fields suppressed, with Gutenberg's added fields updated by this branch.

**Comment by spacedmonkey** [comment]
Related: https://github.com/WordPress/wordpress-develop/pull/264

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (TimothyBJacobs) [inline packages/edit-site/src/components/template-switcher/theme-preview.js:34]: Why isn't this the rendered version? The rendered version has `wptexturize` applied which handles things like curly quotes.

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (TimothyBJacobs) [inline packages/edit-site/src/components/template-switcher/theme-preview.js:17]: Why not `rendered` here as well? We get a safe fallback if the theme doesn't have a proper name listed.

**Comment by TimothyBJacobs** [comment]
I left some comments. I think we also need to consider the value of `_fields` when checking that a field has not been included by Core already.

I am somewhat worried about making sure these stay in sync with Core. Have we ruled out requiring trunk for experimental features like this?

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (ockham) [reply packages/edit-site/src/components/template-switcher/theme-preview.js:34]: My thinking was that the rendered version might contain HTML markup (stuff like `<a />` or `<em />`) and would thus require us to `__dangerouslySetInnerHtml` to render it in React, which I wanted to avoid.

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (ockham) [reply packages/edit-site/src/components/template-switcher/theme-preview.js:17]: ([see](https://github.com/WordPress/gutenberg/pull/23321#discussion_r443795398))

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (TimothyBJacobs) [reply packages/edit-site/src/components/template-switcher/theme-preview.js:34]: It could, but it has `wp_kses` applied to be safe.

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (TimothyBJacobs) [reply packages/edit-site/src/components/template-switcher/theme-preview.js:17]: I think we should, that matches core's behavior see `wp_prepare_themes_for_js`.

**Comment by ockham** [comment]
> I left some comments. I think we also need to consider the value of `_fields` when checking that a field has not been included by Core already.

Ah yeah, good point. I'll modify it to that effect :+1: 

> I am somewhat worried about making sure these stay in sync with Core. Have we ruled out requiring trunk for experimental features like this?

I haven't really been in any conversation regarding that. TBH I'd rather avoid that -- I think it's fair to require a rock-solid implementation of e.g. a REST API endpoint in Core, but I was happy that it wasn't a blocker for developing some code on the client side that depended on it.

Getting a preliminary version of the endpoint additions into Gutenberg allowed me (and others) to iterate quickly on the client-side code, while I was at the same time refining the Core implementation of the endpoint (in accordance with feedback). Overall, it took about 8 weeks until the latter was ready to be merged; that's the amount of time I would've had to wait if we had required those changes to be in Core's trunk for Gutenberg to use it.

**Comment by TimothyBJacobs** [comment]
> Getting a preliminary version of the endpoint additions into Gutenberg allowed me (and others) to iterate quickly on the client-side code, while I was at the same time refining the Core implementation of the endpoint (in accordance with feedback). Overall, it took about 8 weeks until the latter was ready to be merged; that's the amount of time I would've had to wait if we had required those changes to be in Core's trunk for Gutenberg to use it.

I think I did a bad job of explaining what I meant. The idea would still be that code lives in Gutenberg first, but once it is merged to trunk, trunk would be required. That way we don't have to worry about any future changes to trunk making it back to the Gutenberg plugin or the syncing back to Gutenberg from Core was incorrect.

So the development flow would be...

1. Develop API in Gutenberg.
2. Stabilize API and include in trunk.
3. Drop development API in Gutenberg and use the trunk version instead.

**Comment by ockham** [comment]
> > Getting a preliminary version of the endpoint additions into Gutenberg allowed me (and others) to iterate quickly on the client-side code, while I was at the same time refining the Core implementation of the endpoint (in accordance with feedback). Overall, it took about 8 weeks until the latter was ready to be merged; that's the amount of time I would've had to wait if we had required those changes to be in Core's trunk for Gutenberg to use it.
> 
> I think I did a bad job of explaining what I meant. The idea would still be that code lives in Gutenberg first, but once it is merged to trunk, trunk would be required. That way we don't have to worry about any future changes to trunk making it back to the Gutenberg plugin or the syncing back to Gutenberg from Core was incorrect.
> 
> So the development flow would be...
> 
> 1. Develop API in Gutenberg.
> 2. Stabilize API and include in trunk.
> 3. Drop development API in Gutenberg and use the trunk version instead.

Right, thanks for clarifying! But wouldn't that mean that once a new version of the Gutenberg plugin is released after 3., it could no longer be tested on a stable WordPress Core install?

**Comment by youknowriad** [comment]
we still need to maintain compatibility in the Gutenberg plugin with 5.4 even after 5.5 lands.

**Comment by ockham** [comment]
> > I left some comments. I think we also need to consider the value of `_fields` when checking that a field has not been included by Core already.
> 
> Ah yeah, good point. I'll modify it to that effect 

I just tested adding some `_fields` query args to the endpoint, and it seems to work already -- probably because the [`rest_filter_response_fields` filter](https://github.com/WordPress/WordPress/blob/be16bb9fba76592cbe8a0e94c86bca6fd2aa123a/wp-includes/rest-api.php#L745) is [run _after_](https://github.com/WordPress/WordPress/blob/be16bb9fba76592cbe8a0e94c86bca6fd2aa123a/wp-includes/rest-api.php#L190) `gutenberg_filter_rest_prepare_theme`. I'll leave as-is, since I don't think the little extra-optimization for this back-compat patch is worth copy-pasting a bunch of [parsing logic](https://github.com/WordPress/WordPress/blob/be16bb9fba76592cbe8a0e94c86bca6fd2aa123a/wp-includes/rest-api.php#L746-L790).

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (ockham) [reply packages/edit-site/src/components/template-switcher/theme-preview.js:34]: Hmm, okay. I can see that for `name`, but with `description` here, there's an extra complication -- I'm `truncate()`ing, so we might end up with an opening HTML tag that messes up the rest of the markup :confused:

Comment (ockham) [reply packages/edit-site/src/components/template-switcher/theme-preview.js]: ec6f16d6da

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (ockham) [reply packages/edit-site/src/components/template-switcher/theme-preview.js]: ec6f16d6da

**Comment by ockham** [comment]
@TimothyBJacobs I _think_ I addressed your feedback, would you mind giving this another look? :blush: :pray:

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (TimothyBJacobs) [reply packages/edit-site/src/components/template-switcher/theme-preview.js]: It looks like this could be possible with https://github.com/arendjr/text-clipper.

**Comment by TimothyBJacobs** [comment]
> we still need to maintain compatibility in the Gutenberg plugin with 5.4 even after 5.5 lands.

I'll move this to a separate issue, but I think doing this for some things will be completely impossible, like the navigation screen due to how batch requests need to be implemented in Core. I also think its going to prove difficult to keep the larger REST API changes in sync.

> I think I addressed your feedback, would you mind giving this another look?

Looks good. I still wish we'd use the same data ( rendered description ) that we do in wp-admin, but I can see how that complicates things.

**Review by TimothyBJacobs — APPROVED** [review]

File: packages/edit-site/src/components/template-switcher/theme-preview.js
Comment (ockham) [reply packages/edit-site/src/components/template-switcher/theme-preview.js]: Thanks, I wasn't aware of that! I think it's worth considering, but since we'd require an extra dependency for truncating the theme description down to ~120 chars, it might be controversial, so I'd rather not block this PR on it. We can open a separate issue/PR tho!

**Comment by ockham** [comment]
> > we still need to maintain compatibility in the Gutenberg plugin with 5.4 even after 5.5 lands.
> 
> I'll move this to a separate issue, but I think doing this for some things will be completely impossible, like the navigation screen due to how batch requests need to be implemented in Core. I also think its going to prove difficult to keep the larger REST API changes in sync.

I might be missing something here, but is this related to the changes in this PR? The way we're adding fields to the `/themes` endpoint should work with WP 5.4 and 5.5 regardless, shouldn't it? :thinking: (Retaining that sort of compatibility was kind of the point of this PR :sweat_smile: )

> > I think I addressed your feedback, would you mind giving this another look?
> 
> Looks good. I still wish we'd use the same data ( rendered description ) that we do in wp-admin, but I can see how that complicates things.

Thank you! In https://github.com/WordPress/gutenberg/pull/23321#discussion_r446626223, I'm suggesting a follow-up issue/PR to further look into that :slightly_smiling_face:

**Comment by ockham** [comment]
The test setup has changed and tests seem to be stuck. I'll rebase.

**Pull request merged**
